apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1459 @@
+        }
+    }
+    if (CSVRender::PagedWorldspaceWidget *paged =
+        dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
+    {
+        paged->resetAllAlteredHeights();
+        mTotalDiffY = 0;
+    }
+}
+
+void CSVRender::TerrainShapeMode::primarySelectPressed(const WorldspaceHitResult& hit)
+{
+    if(hit.hit && hit.tag == 0)
+    {
+        selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 0, false);
+    }
+}
+
+void CSVRender::TerrainShapeMode::secondarySelectPressed(const WorldspaceHitResult& hit)
+{
+    if(hit.hit && hit.tag == 0)
+    {
+        selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 1, false);
+    }
+}
+
+bool CSVRender::TerrainShapeMode::primaryEditStartDrag (const QPoint& pos)
+{
+    WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+
+    mDragMode = InteractionType_PrimaryEdit;
+
+    if (hit.hit && hit.tag == 0)
+    {
+        mEditingPos = hit.worldPos;
+        mIsEditing = true;
+        if (mShapeEditTool == ShapeEditTool_Flatten)
+            setFlattenToolTargetHeight(hit);
+    }
+
+    return true;
+}
+
+bool CSVRender::TerrainShapeMode::secondaryEditStartDrag (const QPoint& pos)
+{
+    return false;
+}
+
+bool CSVRender::TerrainShapeMode::primarySelectStartDrag (const QPoint& pos)
+{
+    WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+    mDragMode = InteractionType_PrimarySelect;
+    if (!hit.hit || hit.tag != 0)
+    {
+        mDragMode = InteractionType_None;
+        return false;
+    }
+    selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 0, true);
+    return false;
+}
+
+bool CSVRender::TerrainShapeMode::secondarySelectStartDrag (const QPoint& pos)
+{
+    WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+    mDragMode = InteractionType_SecondarySelect;
+    if (!hit.hit || hit.tag != 0)
+    {
+        mDragMode = InteractionType_None;
+        return false;
+    }
+    selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 1, true);
+    return false;
+}
+
+void CSVRender::TerrainShapeMode::drag (const QPoint& pos, int diffX, int diffY, double speedFactor)
+{
+    if (mDragMode == InteractionType_PrimaryEdit)
+    {
+        WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+        std::string cellId = getWorldspaceWidget().getCellId (hit.worldPos);
+        mTotalDiffY += diffY;
+        if (mIsEditing)
+        {
+            if (mShapeEditTool == ShapeEditTool_Drag) editTerrainShapeGrid(CSMWorld::CellCoordinates::toVertexCoords(mEditingPos), true);
+            if (mShapeEditTool != ShapeEditTool_Drag) editTerrainShapeGrid(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), true);
```

Comment:
Not sure what was the intention, but for me first guess why `mShapeEditTool ` is compared twice is that `editTerrainShapeGrid` may change it. `else` might be a better choice here.

Reply:
No specific intention here. It kind of "grew" into that... Changing that is the tiniest of optimizations, so I just didn't bother with it and forgot. I'll change it to `else` along with other cleanup.